ambry-router/src/main/java/com/github/ambry/router/AdaptiveOperationTracker.java

```diff
@@ -244,12 +273,14 @@ public boolean hasNext() {
     @Override
     public void remove() {
       replicaIterator.remove();
-      if (inflightCount >= getCurrentParallelism() && unexpiredRequestSendTimes.size() > 0) {
+      if (!shouldSendRequestWithoutConsideringMetrics() && unexpiredRequestSendTimes.size() > 0) {
         // we are here because oldest request is past due
         Map.Entry<ReplicaId, Pair<Boolean, Long>> oldestEntry = unexpiredRequestSendTimes.entrySet().iterator().next();
-        expiredRequestSendTimes.put(oldestEntry.getKey(), oldestEntry.getValue().getSecond());
-        unexpiredRequestSendTimes.remove(oldestEntry.getKey());
-        pastDueCounter.inc();
+        if (oldestEntry.getValue().getFirst()) {
```

Comment:
Could you explain why this condition was added?

Reply:
double check. If we are in this if statement (link 276) then it means that the oldest entry should already past due and `oldestEntry.getValue().getFirst()` should return true. Putting this if statement here just for sanity check.